Merge ActionTokenStoreProvider and SingleUseObjectProvider

martin-kanis opened the pull request:
Closes #13334

martin-kanis committed:
Merge ActionTokenStoreProvider and SingleUseObjectProvider

hmlnarik reviewed (dismissed):
Thank you for the PR. I'm looking forward to the providers being merged eventually.

There should be no change to the contract of `SingleUseObjectProvider` though. Could you please develop a different approach to it which leverages existing methods rather than adding new ones?

mhajas commented on model/map/src/main/java/org/keycloak/models/map/singleUseObject/MapSingleUseObjectProvider.java:
```diff
@@ -65,121 +65,69 @@ private ActionTokenValueModel singleUseEntityToAdapter(MapSingleUseObjectEntity
-                .compare(ActionTokenValueModel.SearchableFields.ACTION_VERIFICATION_NONCE, ModelCriteriaBuilder.Operator.EQ, actionTokenKey.getActionVerificationNonce().toString());
+    public void put(String key, long lifespanSeconds, Map<String, String> notes) {
+        // try to parse provided key to check if we are storing an action token
+        ActionTokenKeyModel actionTokenKey = DefaultActionTokenKey.from(key);
```
`MapSingleUseObjectProvider` should not need to distinguish between action-token and other objects. Ideally, we should completely remove `DefaultActionTokenKey` and `ActionTokenValueModel` from this class. Would this be possible? 

This way, we could remove all ActionToken-related fields from the storage and retain only `id`, `objectKey` and `notes`. Do you see any problem with this?

martin-kanis replied:
I removed the distinction between action tokens and single-use objects from  `MapSingleUseObjectProvider`. Currently I'm using id for action tokens from upper layers in format which is defined by serialization of an action token. The format is as follow: `encodedUserId.expiration.actionVerificationNonce.actionId` We need to agree on the format.
This allowed me to remove `userId`, `actionId`, `actionVerificationNonce` fields from everywhere.

Now question is whether we can go even further and remove also `objectKey` and all custom `SingleUseObjectModelCriteriaBuilder|SingleUseObjectKeycloakTransaction|SingleUseObject*Storage`?

mhajas replied:
Thank you @martin-kanis, it looks much cleaner now, and I would say it makes sense to have it this way. 

> Now question is whether we can go even further and remove also objectKey and all custom SingleUseObjectModelCriteriaBuilder|SingleUseObjectKeycloakTransaction|SingleUseObject*Storage?

We can't get rid of `objectKey` completely because JPA is using it. On the other hand, we could move the field to JPA only. Currently, it seems we are having a field, necessary only in JPA available to all other implementations. This probably doesn't need to be solved here, I would prefer to discuss and possibly resolve within another issue. See my other comments for more details.

mhajas reviewed (commented):
@martin-kanis thanks for the changes. Could you please look at my comment and let me know what do you think about it? If you want we can meet and discuss it.

mhajas commented on model/map/src/main/java/org/keycloak/models/map/singleUseObject/MapSingleUseObjectProvider.java:
```diff
@@ -42,7 +40,7 @@
  * @author <a href="mailto:mkanis@redhat.com">Martin Kanis</a>
  */
-public class MapSingleUseObjectProvider implements ActionTokenStoreProvider, SingleUseObjectProvider {
+public class MapSingleUseObjectProvider implements SingleUseObjectProvider {
```
This file still mentions ActionTokens. I will add some ideas that may be added if you like it. 

- It seems we can now remove `singleUseEntityToAdapter` method
- `actionTokenStoreTx` should be renamed to something like `singleUseObjectTx`
- The usage of `ActionTokenValueModel` is a little bit weird now. Should we rename it to `SingleUseObjectModel`? 
- Also, the presence of `ObjectKey` there is weird. For example, in HotRod it is not necessary at all, so it sounds like an implementation detail for JPA so maybe it should be present only there? WDYT? Should we create a new issue for that?

martin-kanis replied:
Thanks. I addressed these suggestions. I renamed both `ActionTokenValueModel` and `ActionTokenKeyModel` and some related classes and variables. 
I agree regarding `ObjectKey` that it is weird and should handle it in a separate issue.

mhajas replied:
I created an issue: [#13989](https://github.com/keycloak/keycloak/issues/13989)

sguilhen reviewed (commented):
This looks great, I like how it simplifies things. I also agree with the `objectKey` comments and it is indeed something only JPA needs to keep the single use objects table consistent with the other tables (i.e. have an Id of type UUID)

mhajas reviewed (commented):
@martin-kanis After a quick discussion with @hmlnarik we agreed it is probably worth refactoring also the legacy store to use the same approach for both single-use and action tokens. Do you know about something that would suggest otherwise?

martin-kanis committed:
Simplify InfinispanSingleUseObjectProvider (use string keys for action tokens, remove remote cache for single-use-objects)

martin-kanis commented:
/rerun

mhajas reviewed (commented):
@martin-kanis Previous implementation of SUOProvider was using this when initializing caches:

https://github.com/keycloak/keycloak/blob/fb24c86a3bf4dcfd661e11288686afe8a88f10ef/model/infinispan/src/main/java/org/keycloak/models/sessions/infinispan/InfinispanSingleUseObjectProviderFactory.java#L68-L79

This is missing from the current implementation. Should we add this? Does this break cross-dc support for SingleUseObjects?

martin-kanis commented:
> @martin-kanis Previous implementation of SUOProvider was using this when initializing caches:
> 
> https://github.com/keycloak/keycloak/blob/fb24c86a3bf4dcfd661e11288686afe8a88f10ef/model/infinispan/src/main/java/org/keycloak/models/sessions/infinispan/InfinispanSingleUseObjectProviderFactory.java#L68-L79
> 
> This is missing from the current implementation. Should we add this? Does this break cross-dc support for SingleUseObjects?

I added a cross-dc test scenario that showed the initialization you mentioned should be returned back. When deleting an object in one node, nodes at the other site still see the object when they request it via the provider (`singleUseStore.get(key)`).

mhajas reviewed (changes requested):
@martin-kanis Few more things around configuration I missed during my review:
- we need to also reflect this merge in Quarkus configuration: https://github.com/keycloak/keycloak/blob/main/quarkus/config-api/src/main/java/org/keycloak/config/StorageOptions.java#L220-L230
- We can cleanup properties definition in pom files `keycloak.actionToken.provider` and `keycloak.actionToken.map.storage.provider`

Also, I executed the pipeline for this change so we are more confident in legacy changes.

mhajas commented:
@martin-kanis here is the Jenkins run: https://master-jenkins.redhat.com/job/universal-test-pipeline-server/2243/ I haven't found any issue caused by this PR, so it seems after addressing my latest suggestions we should be good to go.

martin-kanis commented:
@mhajas Thank you for the comment and running pipelines. I updated the PR and also sent a docs PR https://github.com/keycloak/keycloak-documentation/pull/1697 that fixes mentions of now removed `ActionTokenSpi`.

mhajas reviewed (dismissed):
Thanks @martin-kanis, LGTM!

hmlnarik commented:
@martin-kanis @mhajas Quick question - do quickstarts work with the proposed changes?

hmlnarik commented on model/infinispan/src/main/java/org/keycloak/models/sessions/infinispan/InfinispanSingleUseObjectProviderFactory.java:
```diff
@@ -86,7 +71,15 @@ public void init(Config.Scope config) {
-
+        // It is necessary to put the cache initialization here, otherwise the cache would be initialized lazily, that
+        // means also listeners will start only after first cache initialization - that would be too late
+        if (singleUseObjectCache == null) {
+            synchronized (this) {
+                if (singleUseObjectCache == null) {
```
This synchronization can be removed since `postInit` is executed only once at the initialization in a single thread, so contrary to lazy init, there is only a single caller.

martin-kanis replied:
Removed

hmlnarik reviewed (commented):
Overall looks good to me but I'd like to have @sguilhen / @vramik / @ahus1 's review for the JPA changes before merging

martin-kanis commented:
> @martin-kanis @mhajas Quick question - do quickstarts work with the proposed changes?

I failed to tried it because those 2 QS rely on Keycloak running on Wildfly (see https://github.com/keycloak/keycloak-quickstarts/blob/latest/action-token-required-action/pom.xml#L111-L128). However it should work because the quickstarts implement `actionTokenHandler` SPI that I didn't touch in this PR.

sguilhen reviewed (commented):
From a JPA perspective the changes look good. I just think we might also have to drop the index involving the action verification nonce from the schema:

https://github.com/keycloak/keycloak/blob/8ea3f30d827450d3fd72e3660ab0aabaec6daf46/model/map-jpa/src/main/resources/META-INF/single-use-objects/jpa-single-use-objects-changelog-1.xml#L50-L52

vramik commented on model/map-jpa/src/main/resources/META-INF/single-use-objects/jpa-single-use-objects-changelog-1.xml:
```diff
@@ -47,9 +47,6 @@ limitations under the License.
         <createIndex tableName="kc_single_use_obj" indexName="single_use_obj_expiration">
             <column name="expiration"/>
         </createIndex>
-        <ext:createJsonIndex tableName="kc_single_use_obj" indexName="kc_single_use_obj_nonce">
-            <ext:column jsonColumn="metadata" jsonProperty="fActionVerificationNonce"/>
-        </ext:createJsonIndex>
```
I think this drop of the index has to go to different `changeSet`. Otherwise it would fail with validation when checking the checksums when running in no-downtime mode.

martin-kanis replied:
Thanks. I fixed it. It should work as expected:
```
keycloak=# \d kc_single_use_obj
...
Indexes:
    "kc_single_use_obj_pkey" PRIMARY KEY, btree (id)
    "single_use_obj_entityVersion" btree (entityversion)
    "single_use_obj_expiration" btree (expiration)
    "single_use_obj_objectKey" btree (objectkey)
Referenced by:
    ...
```
There is no longer the following index: `"kc_single_use_obj_nonce" gin ((metadata -> 'fActionVerificationNonce'::text) jsonb_path_ops)`

vramik replied:
Thank you. still it should be reverted to pass the validation.

vramik reviewed (changes requested):
Thanks @martin-kanis. It LGTM, I have just one comment, please see inline.

vramik commented on model/map-jpa/src/main/resources/META-INF/single-use-objects/jpa-single-use-objects-changelog-2.xml:
```diff
@@ -0,0 +1,31 @@
+                   xsi:schemaLocation="http://www.liquibase.org/xml/ns/dbchangelog http://www.liquibase.org/xml/ns/dbchangelog/dbchangelog-3.1.xsd
+                    http://www.liquibase.org/xml/ns/dbchangelog-ext http://www.liquibase.org/xml/ns/dbchangelog/dbchangelog-ext.xsd">
+
+    <changeSet author="keycloak" id="single-use-object-2">
+
+        <dropIndex tableName="kc_single_use_obj" indexName="kc_single_use_obj_nonce"/>
```
Please move the `changeSet` to `jpa-single-use-objects-changelog-1.xml`. The number should correspond to current_schema_version and it remains the same.

martin-kanis replied:
fixed

vramik reviewed (approved):
Thank you @martin-kanis. From the jpa point of view it seems good to me.

mhajas reviewed (approved):
Thanks @martin-kanis
